cms/apps/media/models.py

```diff
@@ -212,10 +318,103 @@ class Video(models.Model):
         null=True
     )
 
+    external_video = models.CharField(
+        max_length=255,
+        blank=True,
+        null=True,
+        help_text='Please provide a youtube.com or vimeo.com URL',
+    )
+
+    # Secret fields for external videos - populated from the URL when the form is saved.
+    external_video_iframe_url = models.TextField(
+        null=True,
+        blank=True,
+    )
+
+    external_video_id = models.CharField(
+        max_length=32,
+        blank=True,
+        null=True,
+    )
+
+    external_video_service = models.CharField(
+        max_length=32,
+        blank=True,
+        null=True,
+    )
+    # End secret fields
+
     def __str__(self):
         """Returns the title of the media."""
         return self.title
 
+    def clean(self):
+        if (self.high_resolution_mp4 or self.low_resolution_mp4) and self.external_video:
+            raise ValidationError({
+                'high_resolution_mp4': "Please provide either a locally hosted file or an external file, not both."
+            })
+
+        if self.external_video:
+            info = get_video_info(self.external_video)
+            if info:
+                self.external_video_iframe_url = info['src']
+
+                if not self.external_video_iframe_url:
+
+                    raise ValidationError({
+                        'external_video': "Couldn't determine how to embed this video. Maybe the video's privacy settings disallow embedding?"
+                    })
+                self.external_video_id = info['id']
+                self.external_video_service = info['service']
+
+        return super().clean()
+
+    def embed_html(self, loop=False, autoplay=False, controls=False, mute=False, youtube_parameters=None, width=640, height=360):
```

Comment:
The width and height of the embed should be controlled in CSS rather than HTML width and height attributes.